File: databases/slatedb/python/src/recipe_kv_store.py
"""
Recipe: Basic KV Store Operations
Database: SlateDB
Description: Demonstrates opening a SlateDB database backed by an in-memory
             object store, putting, getting, deleting, and listing keys.

Usage: python src/recipe_kv_store.py
"""

import asyncio
import tempfile
from pathlib import Path

from slatedb.uniffi._slatedb_uniffi import DbBuilder, KeyRange, ObjectStore


async def main():
    # 1. Setup -- create a temporary directory for the local object store
    tmpdir = tempfile.mkdtemp(prefix="slatedb_recipe_")
    store = ObjectStore.resolve(f"file://{tmpdir}")
    builder = DbBuilder("/recipe-db", store)
    db = await builder.build()

    # 2. Write -- insert key-value pairs
    await db.put(b"name", b"SlateDB")
    await db.put(b"type", b"key-value")
    await db.put(b"year", b"2024")
    await db.put(b"license", b"Apache-2.0")
    print("Inserted 4 key-value pairs.")

    # 3. Read -- retrieve individual keys
    name = await db.get(b"name")
    kind = await db.get(b"type")
    print(f"name = {name}")
    print(f"type = {kind}")

    # 4. Delete -- remove a key
    await db.delete(b"license")
    deleted = await db.get(b"license")
    print(f"license (after delete) = {deleted}")

    # 5. List -- iterate over all keys
    print("\nAll entries:")
    key_range = KeyRange(start=None, start_inclusive=False, end=None, end_inclusive=False)
    iterator = await db.scan(key_range)
    while (kv := await iterator.next()) is not None:
        print(f"  {kv.key} => {kv.value}")

    # 6. Cleanup
    await db.shutdown()
    import shutil
    shutil.rmtree(tmpdir, ignore_errors=True)
    print(f"\nCleaned up {tmpdir}")


if __name__ == "__main__":
    asyncio.run(main())

File: databases/redb/rust/src/bin/recipe_kv_store.rs
//! Recipe: Basic KV Store Operations
//! Database: redb
//! Description: Demonstrates opening a database, inserting, getting, deleting
//!     keys, and performing a cursor-based iteration using the redb typed table API.
//!
//! Usage: cargo run --bin recipe_kv_store

use redb::{Database, ReadableDatabase, ReadableTable, TableDefinition};

const TABLE: TableDefinition<&str, &str> = TableDefinition::new("kv");

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let dir = tempfile::tempdir()?;
    let db_path = dir.path().join("demo.redb");

    // 1. Setup -- create the database
    let db = Database::create(&db_path)?;

    // 2. Write -- insert key-value pairs in a write transaction
    let write_txn = db.begin_write()?;
    {
        let mut table = write_txn.open_table(TABLE)?;
        table.insert("name", "redb")?;
        table.insert("type", "key-value")?;
        table.insert("year", "2021")?;
        table.insert("license", "Apache-2.0 OR MIT")?;
    }
    write_txn.commit()?;

    // 3. Read -- retrieve individual keys in a read transaction
    let read_txn = db.begin_read()?;
    let table = read_txn.open_table(TABLE)?;

    match table.get("name")? {
        Some(entry) => println!("name = {}", entry.value()),
        None => println!("name = not found"),
    }
    match table.get("type")? {
        Some(entry) => println!("type = {}", entry.value()),
        None => println!("type = not found"),
    }

    // 4. Delete -- remove a key in a write transaction
    let write_txn = db.begin_write()?;
    {
        let mut table = write_txn.open_table(TABLE)?;
        table.remove("license")?;
    }
    write_txn.commit()?;

    let read_txn = db.begin_read()?;
    let table = read_txn.open_table(TABLE)?;
    match table.get("license")? {
        Some(entry) => println!("license after delete = {}", entry.value()),
        None => println!("license after delete = None"),
    }

    // 5. Scan -- iterate over all entries
    println!("\nAll entries:");
    let read_txn = db.begin_read()?;
    let table = read_txn.open_table(TABLE)?;
    let iter = table.iter()?;
    for result in iter {
        let (key, value) = result?;
        println!("  {} => {}", key.value(), value.value());
    }

    println!("\nDone.");
    Ok(())
}
